src/cgame/cg_rocket_draw.cpp

```diff
@@ -2174,6 +2174,131 @@ class PredictedMineEfficiencyElement : public HudElement
 	std::unordered_map<int, std::string> pluralSuffix;
 };
 
+class BarbsHudElement : public HudElement
+{
+public:
+	BarbsHudElement ( const Rocket::Core::String& tag ) :
+	HudElement ( tag, ELEMENT_ALIENS ),
+	numBarbs( 0 ),
+	maxBarbs( 0 ),
+	regenerationInterval ( 0 ),
+	t0 ( 0 ),
+	offset ( 0 ) {}
+
+	void OnAttributeChange( const Rocket::Core::AttributeNameList& changed_attributes )
+	{
+		HudElement::OnAttributeChange( changed_attributes );
+		if ( changed_attributes.find( "src" ) != changed_attributes.end() )
+		{
+			maxBarbs = BG_Weapon( WP_ALEVEL3_UPG )->maxAmmo;
+			if ( maxBarbs > 0 )
+			{
+				src = GetAttribute<Rocket::Core::String>( "src", "" );
+				Rocket::Core::String base( va("<img class='barbs' src='%s' />", src.CString() ) );
+				Rocket::Core::String rml;
+
+				for ( int i = 0; i < maxBarbs; i++ )
+				{
+					rml += base;
+				}
+
+				// SetInnerRML deletes all child Elements => bring children up to date
+				children.clear();
+				SetInnerRML( rml );
+				for( int i = 0; i < maxBarbs; i++) {
+					children.push_back( GetChild( i ) );
+				}
+			}
+			else
+			{
+				children.clear();
+				SetInnerRML( "" );
+			}
+		}
+	}
+
+	void DoOnUpdate()
+	{
+		int barbs = cg.snap->ps.ammo;
+		int interval = GetInterval();
+
+		// start regenerating barb now
+		if ( barbs > numBarbs || ( barbs < numBarbs && numBarbs == maxBarbs ) )
+		{
+			t0 = cg.time;
+			// sin(-pi/2) is minimal
+			offset = -M_PI_2;
+		}
+		// change regeneration speed
+		else if ( interval != regenerationInterval )
+		{
+			// black math magic to avoid sudden jumps in opacity
+			t0 = cg.time;
+			offset = asin( GetSin() ) - M_PI_2;
+			regenerationInterval = interval;
+		}
+		numBarbs = barbs;
+
+		for ( int i = 0; i < children.size(); i++)
+		{
+			if (i < numBarbs ) // draw existing barbs
+			{
+				children[i]->SetProperty("color", "white");
+				children[i]->SetProperty("opacity", "1.0");
+			}
+			else if (i == numBarbs ) // draw regenerating barb
+			{
+				float opacity = GetSin() / 3 + 0.5; // normalize
```

Comment:
I see. I'd just remove the comment then as normalizing kind of implies bringing something to a scale of [0,1]. Also I do find, even with you currently using grey instead of white, that the opacity gets too high (currently 5/6). I would have imagined something like a range of 0.25 to 0.5, hence `GetSin() / 8.0 + (3.0/8.0)`.